DEMO-147: Added RichText Custom Tags: ezyoutube, eztwitter, ezfacebook

alongosz opened the pull request:
JIRA issue: https://jira.ez.no/browse/DEMO-147

This PR adds sample Custom Tags.

**TODO**:
- [x] Add `ezyoutube` Custom Tag configuration, Admin UI icon, translations and frontend template.
- [x] Add `eztwitter` Custom Tag configuration, Admin UI icon, translations and frontend template.
- [x] Add `ezfacebook` Custom Tag configuration, Admin UI icon, translations and frontend template.
- [ ] Improve `ezfacebook` icon
- [ ] Send for review
- [ ] Test by QA

bdunogier commented:
Looks good overall. One remark about theming though: since the templated are part of the app's design, the templates won't be found when rendering content in the admin. One solution would be to make them part of the standard design, used by both front and backend.

alongosz commented:
> since the templated are part of the app's design, the templates won't be found when rendering content in the admin. One solution would be to make them part of the standard design, used by both front and backend.

@bdunogier 
True, but that's for the future when changes introducing standard design are gonna be added to meta.
Right now `ezplatform-demo` uses `main` design and `tastefulplanet` theme for both front and backend. So the templates are also rendered nicely in the admin.

bdunogier commented:
Oh, okay. Great then :)

alongosz commented:
also review ping @dew326 (can't do it via review request).

sunpietro commented on app/Resources/views/themes/tastefulplanet/fields/ezrichtext/custom_tags/ezfacebook.html.twig:
```diff
@@ -0,0 +1,14 @@
+    }(document, 'script', 'facebook-jssdk'));
+</script>
+
+<div
```
I prefer this way: 
```twig
<div 
    class="fb-post" 
    style="display: {{ params.display }};" 
    data-href="{{ params.post_url }}" 
    data-width="{{ params.width ? params.width : '' }}">
</div>
```

alongosz replied:
Fixed in 4014bf5.

alongosz replied:
And amended in 5b17535. Hope this time It's actually well formed.

sunpietro commented on app/Resources/views/themes/tastefulplanet/fields/ezrichtext/custom_tags/ezfacebook.html.twig:
```diff
@@ -0,0 +1,14 @@
+<script>(function(d, s, id) {
```
CS: Please, move it below HTML code.

alongosz replied:
Fixed in 4014bf5.

sunpietro commented on app/Resources/views/themes/tastefulplanet/fields/ezrichtext/custom_tags/eztwitter.html.twig:
```diff
@@ -0,0 +1,41 @@
+<script>window.twttr = (function(d, s, id) {
```
Please, replace this code with following one:
```javascript
<script>
(function (global, doc, twttr) {
    twttr = twttr || (function(d, s, id) {
        var js, 
            fjs = d.getElementsByTagName(s)[0],
            t = global.twttr || {};
        if (d.getElementById(id)) return t;
        js = d.createElement(s);
        js.id = id;
        js.src = 'https://platform.twitter.com/widgets.js';
        fjs.parentNode.insertBefore(js, fjs);

        t._e = [];
        t.ready = function(f) {
            t._e.push(f);
        };

        return t;
    }(doc, 'script', 'twitter-wjs'));

    const tweetUrl = '{{ params.tweet_url }}';
    const tweetUrlParts = tweetUrl.split('/');
    const tweetId = tweetUrlParts[tweetUrlParts.length - 1];

    twttr.ready(function () {
        twttr.widgets.load(
            twttr.widgets.createTweet(
                tweetId,
                document.querySelector('.tweet-container[data-url="{{ params.tweet_url }}"]'),
                {
                    'cards': '{{ params.cards is defined ? params.cards : '' }}',
                    'conversation': '{{ params.conversation is defined ? params.conversation : '' }}',
                    'theme': '{{ params.theme is defined ? params.theme : '' }}',
                    'linkColor': '{{ params.link_color is defined ? params.link_color : '' }}',
                    'width': '{{ params.width is defined ? params.width : '' }}',
                    'align': '{{ params.align is defined ? params.align : '' }}',
                    'lang': '{{ params.lang is defined ? params.lang : '' }}',
                    'dnt': '{{ params.dnt is defined ? params.dnt : '' }}'
                }
            )
        );
    });
})(window, window.document, window.twttr);
</script>
<div class="tweet-container" data-url="{{ params.tweet_url }}"></div>

```

alongosz replied:
4014bf5 contains the fix according to recent internal discussion due to other issues discovered.

inakijv commented:
Twitter and Youtube are already included. Please check for: `#twitter` & `#video`
I am going to work on the Facebook one.
Besides, I agree @SylvainGuittard, due to color contrast standards we are following we should not use colors for standalone icons, we should use it when signifying actions or changes of state.

dew326 commented on app/Resources/views/themes/tastefulplanet/fields/ezrichtext/custom_tags/ezfacebook.html.twig:
```diff
@@ -0,0 +1,14 @@
+<div
+        class="fb-post" style="display: {{ params.display }};"
+        data-href="{{ params.post_url }}"
+        data-width="{{ params.width ? params.width : '' }}"
```
Did you check the issue I found in ezplatform-ee-demo with empty width?

alongosz replied:
Yes, addressed in [5b17535](https://github.com/ezsystems/ezplatform-demo/commit/5b17535465ddc9cf75d4baec624b7458caf31c7f#diff-518cad18e7ba0a166286fa1fa2a5b683R5), my mistake.

alongosz commented:
Improved `ezyoutube.html.twig` in 6851b5d according to @sunpietro suggestions.

Recent fixes also include:
- Facebook svg provided by @inakijv (we had to adjust it a little bit because it didn't show on toolbar in Online Editor)
- Modified paths pointing to existing svg's for `ezyoutube` and `eztwitter`
- Changed default with of Twitter post from 250 to 500 (it looks better IMHO)

Any more changes needed (_speak now or forever hold your peace_ :grinning: )?
// cc @SylvainGuittard @inakijv @sunpietro @dew326 @damianz5

alongosz commented:
Squashed commits after review and sent to QA, so they can have fun with that also ;)

bdunogier commented on app/config/custom_tags.yml:
```diff
@@ -0,0 +1,87 @@
+ezpublish:
+    system:
+        admin_group:
```
Is that correct ? Or is the frontend also part of the `admin_group` ?
I expected that a tag needs to be enabled for both frontend and backend.

alongosz replied:
> I expected that a tag needs to be enabled for both frontend and backend.

Actually... no. It's about being able to add tag in Online Editor. What was saved has to be rendered on frontend. Let's imagine a case where someone decides to no longer support some tag. In Online Editor it will result in validation error on save, which is IMHO fine. However if we added such checking on frontend it would also make sense to validate it. Such thing would result in fronted site crashing. 
So I'm not a huge fan of the idea.

But... We could introduce that for frontend by saying "let's just silently not render it if it's not enabled". The problem I have here is Product stability. This change would be a BC break after `2.1`. We still have some time until stable 2.1, but I think it's too little to do it.

bdunogier replied:
I see. Let's not break anything :)

The priority is to make sure the Admin won't break if there's a custom tag, whatever the reason, in a richtext field (after a migration for instance). It could throw a warning though.

Related to that, what about validation of custom tag attributes on publishing ? We didn't have that feature before. Isn't there a possibility that content that was valid before gets flagged as invalid ? If it is the case, we may want to make validation optional, disable it by default, and notify that having it to false is deprecated, and the default will be changed in the next major version.

We can think of the same procedure as for enabled/disabled tags in the context of rendering.

alongosz replied:
You've touched interesting subject, let's talk about it internally.

SylvainGuittard commented on app/Resources/views/themes/tastefulplanet/fields/ezrichtext/custom_tags/ezyoutube.html.twig:
```diff
@@ -0,0 +1,9 @@
+        type="text/html"
+        width="{{ params.width }}"
+        height="{{ params.height }}"
+        src="{{ params.video_url|replace({'https://youtu.be/' : 'https://www.youtube.com/embed/'}) }}?autoplay={{ params.autoplay == 'true' ? 1 : 0 }}"
```
I just tried the PR. `URL` is kind of confusing, because I copied / pasted the URL of the browser, not the one with the `SHARE` button.
We can maybe change `Video URL` by `Video ID`. What do you think?

alongosz replied:
It just requires some doc. We've discussed Video ID vs some URL and came to conclusion that an average user won't understand what is video ID and also won't be able to select properly ID portion to be copied. Unfortunately it's not possible to embed video copying direct URL.

You have "Share" button below video which contains valid URL:
![image](https://user-images.githubusercontent.com/7099219/37620451-8e497e96-2bbc-11e8-9988-7d6b07159395.png)

But yes, if you believe Video ID is better - we can change it, after all Custom Tags are... custom :)

SylvainGuittard replied:
I totally understand the approach. Let's keep it like that. :)
Thanks @alongosz

bdunogier replied:
By the way, at some point, I remember thinking of an extra label for attributes, that would be filled using translations, like `custom_tags.foo.attributes.bar.helptext`. For such cases, it would allow us to document this item a little bit in the dialog.

SylvainGuittard commented on app/config/custom_tags.yml:
```diff
@@ -0,0 +1,87 @@
+        # Define Custom Tags
+        custom_tags:
+            ezyoutube:
+                template: '@ezdesign/fields/ezrichtext/custom_tags/ezyoutube.html.twig'
```
Is this template supposed to be used on frontend and / or backend?

alongosz replied:
It's frontend, but backoffice when previewing (not editing) content uses frontend templates. This is a perfect example showing the power of eZ Design Engine. 

1. Create separate design for frontend and separate for backend with separate themes.
e.g.:
```yml
ezpublish:
    system:
        site_group:
            design: app
        admin_group:
            design: admin
design_list:
    app: [app, tastefulplanet]
    admin: [admin, tastefulplanet]
```

2. Then create the following templates for front and admin respectively:
- `app/Resources/views/themes/app/fields/ezrichtext/custom_tags/ezyoutube.html.twig`
- `app/Resources/views/themes/admin/fields/ezrichtext/custom_tags/ezyoutube.html.twig`

3. Remove `ezyoutube.html.twig` from `tastefulplanet` theme.
4. Clear Symfony cache.

You should get different template on frontend and different in Admin UI view.

SylvainGuittard replied:
Indeed, it's using the same template for frontend and backend for `content / view`. I tested it and it works.
My initial question was for `content / edit`.

alongosz replied:
> My initial question was for `content / edit`.

`content / edit` preview needs some work both for backend and frontend (js) part. AFAIR this was supposed to be a separate effort to be scheduled in the future. 
// cc @bdunogier

SylvainGuittard replied:
Thanks @alongosz for clarifying this point

SylvainGuittard commented on app/Resources/views/themes/tastefulplanet/fields/ezrichtext/custom_tags/ezyoutube.html.twig:
```diff
@@ -0,0 +1,9 @@
+<div{% if params.align is defined %} style="text-align: {{ params.align }};"{% endif %}>
```
`params.align`: Is there a way to replace this value by the value of the alignment buttons?
![image](https://user-images.githubusercontent.com/1157233/37620606-1f3ecd72-2b93-11e8-980d-c286c969038c.png)

alongosz replied:
ping @dew326

SylvainGuittard replied:
If it's too complicated for this release, we should remove the buttons and not confuse users.

alongosz replied:
@SylvainGuittard I've established with @dew326 that a proper value of alignment is set and saved to a database, but RichText Custom Tag renderer doesn't pass it to a template, so it's a Kernel bug :scream: 

Reported as [EZP-28972](https://jira.ez.no/browse/EZP-28972).

dew326 replied:
There are not bugs in Kernel, it's only a challenge.

sunpietro replied:
![not-bug-its-feature](https://user-images.githubusercontent.com/1736658/37643814-0f0361d6-2c22-11e8-819c-55f44eabb8b3.jpg)

SylvainGuittard replied:
Can we have this issue ([EZP-28972](https://jira.ez.no/browse/EZP-28972)) fixed for v2.1.0?

alongosz replied:
Yes, working on it.

alongosz commented:
@SylvainGuittard @damianz5 I've completed ezsystems/ezpublish-kernel#2286 and here I applied `align` variable in all the templates and removed unnecessary attributes.
We probably need QA on this as I've mentioned earlier. // cc @micszo

micszo commented:
@alongosz I'm on it.
